python_modules/dagster/dagster/_config/pythonic_config/pydantic_compat_layer.py

```diff
@@ -0,0 +1,137 @@
+from typing import (
+    Dict,
+    Type,
+)
+
+import pydantic
+
+PydanticUndefined = None
+try:
+    from pydantic_core import PydanticUndefined as _PydanticUndefined  # type: ignore
+
+    PydanticUndefined = _PydanticUndefined
+except:
+    pass
+
+
+from pydantic import BaseModel
+
+from .attach_other_object_to_context import (
+    IAttachDifferentObjectToOpContext as IAttachDifferentObjectToOpContext,
+)
+
+USING_PYDANTIC_2 = int(pydantic.__version__.split(".")[0]) >= 2
+
+
+class ModelFieldCompat:
+    """Wraps a Pydantic model field to provide a consistent interface for accessing
+    metadata and annotations between Pydantic 1 and 2.
+    """
+
+    def __init__(self, field):
+        self.field = field
+
+    @property
+    def annotation(self):
+        return self.field.annotation
+
+    @property
+    def metadata(self):
+        return getattr(self.field, "metadata", None)
+
+    @property
+    def alias(self):
+        return self.field.alias
+
+    @property
+    def default(self):
+        return self.field.default
+
+    @property
+    def description(self):
+        # Pydantic 1.x
+        field_info = getattr(self.field, "field_info", None)
+        if field_info:
+            return field_info.description
+
+        # Pydantic 2.x
+        return getattr(self.field, "description", None)
+
+    def is_required(self):
+        # Pydantic 2.x
+        if hasattr(self.field, "is_required"):
+            return self.field.is_required()
+
+        # Pydantic 1.x
+        return self.field.required
+
+    @property
+    def discriminator(self):
+        # Pydantic 2.x
+        if hasattr(self.field, "discriminator"):
+            return self.field.discriminator
+
+        # Pydantic 1.x
+        return getattr(self.field, "discriminator_key", None)
```

Comment:
should we just explicitly have different codepaths for each version here? I think that will be more bulletproof

Reply:
Yes, that's a better approach I think. Reworked.